# Τεχνητή Νοημοσύνη: Εργαστηριακή Άσκηση 2
---

Ο στόχος της εργασίας είναι η κατασκευή ενός συστήματος προτάσεων (Recommendation System) για ταινίες. Οι προτάσεις αυτές θα πηγάζουν τόσο από τα χαρακτηριστικά της ταινίας όσο και από ορισμένες αξιολογήσεις του κάθε χρήστη.

Στα δεδομένα της άσκησης περιλαμβάνονται ένα αρχείο με το όνομα movies_metadata.csv το όποιο περιέχει τα χαρακτηριστικά κάθε ταινίας όπως θέμα, σκηνοθέτης ηθοποιοί, λέξεις κλειδιά κ.α. από το imdb καθώς και τα αρχεία ratings.csv τα όποια περιέχουν πραγματικές αξιολογήσεις χρηστών, χωρισμένες σε train και σε test.


# Εκφώνηση
Στην παρούσα εργασία σας ζητείται να μελετήσετε και να υλοποιήσετε τα παρακάτω:<br>

### Μέρος 1
Στο μέρος 1 και 2 θα εργαστείτε μόνο με το αρχείο movies_metadata.csv, ενώ στο μέρος τρία θα δουλέψετε και με τα αρχεία των αξιολογήσεων.
#### Ερώτημα 1α

Αρχικά θα πρέπει, αφού μελετήσετε τη δομή και τα χαρακτηριστικά του movies_metadata.csv, να κατασκευάσετε μια βάση γνώσης για την Prolog η όποια ουσιαστικά θα αποτελεί τον κόσμο με τον όποιο θα εργαστείτε στην συνέχεια. Τα κατηγορήματα που θα δημιουργηθούν θα σας βοηθήσουν και στην κατασκευή του recommender και θα είναι της μορφής:

```
director(Movie, Director).
genre(Movie, Genre).
```
#### Ερώτημα 1β

Αφού δημιουργήσετε τον κόσμο του προβλήματος, στη συνέχεια καλείστε να δημιουργήσετε, σε Prolog, απλούς κανόνες οι οποίοι θα βρίσκουν όλες τις ταινίες με:
1.	Κοινό θέμα (κάποιες λέξεις σχετικά με το genre κοινές)
2.	Αρκετά κοινό θέμα (κάποιες λιγότερες λέξεις σχετικά με το genre κοινές π.χ. 3)
3.	Σχετικά κοινό θέμα (λίγες λέξεις σχετικές με το genre κοινές π.χ. 1)
4.	Κοινός σκηνοθέτης
5.	Ακριβώς ίδια πλοκή (κάποιες λέξεις κλειδιά της πλοκής κοινές)
6.	Σχετικά ίδια πλοκή (κάποιες λιγότερες λέξεις κλειδιά κοινές)
7.	Ίδιους τους βασικούς ηθοποιούς (και τους 3)
8.	Αρκετά ίδιους βασικούς ηθοποιούς (ορισμένους βασικούς ηθοποιούς κοινούς π.χ. 2)
9.	Σχετικά ίδιους ηθοποιούς (π.χ. 1 από τους 3)
10.	Ίδια γλώσσα
11.	Είναι έγχρωμες ή ασπρόμαυρες
12. Κοινό studio παραγωγής
13. Κοινή χώρα παραγωγής
14. Ίδια δεκαετία

Αξίζει να σημειωθεί ότι στα παραπάνω μπορείτε να προσθέσετε περισσότερα ερωτήματα ή να αλλάξετε την κλιμακωσιμότητα των queries (πέρα από το ίδιο, αρκετά ίδιο, σχετικά ίδιο) που θα κατασκευάσετε(π.χ. μια κλίμακα από 1 στα 5 όπου αυτό είναι δυνατόν), μιας και αυτά στην συνέχεια θα χρησιμοποιηθούν για την λειτουργία του recommender. Έτσι μπορείτε να προσθέσετε queries με τα όποια μπορεί να παράγονται καλύτερες συστάσεις  (το αρχείο movie_metadata.csv περιέχει πολλές πληροφορίες ακόμα για κάθε ταινία όπως έτος κυκλοφορίας, βαθμολογία στο imdb, facebοok_likes κ.α.). Το συγκεκριμένο μέρος εργασίας είναι προπαρασκευαστικό οπότε όσο καλύτερα και πλουσιότερα τα ερωτήματα που θα φτιάξετε σε αυτό το μέρος τόσο καλύτερη θα είναι η απόδοσή των συστημάτων συστάσεων των επόμενων ερωτημάτων.

Περισσότερες πληροφορίες για το dataset μπορείτε να διαβάσετε σε αυτό το [Link](https://www.kaggle.com/georgefila/movies-metadata).
### Μέρος 2: Recommendation System
Στο σημείο αυτό καλείστε με βάση αυτά που κάνατε στο μέρος 1 να κατασκευάσετε queries τα όποια θα σας επιστρέφουν παρόμοιες (σε χαρακτηριστικά) ταινίες. Τα ερωτήματα αυτά θα είναι κλιμακούμενα, δηλαδή θα υπάρχουν ερωτήματα που επιστρέφουν αρκετά κοινές ταινίες αλλά και που επιστρέφουν λιγότερο και λιγότερο κοινές (σε μια κλίμακα π.χ. από 1 σε 5). Για παράδειγμα:

```
find_simmilar_movies_5("Pirates Of The Caribbean", M).
M = "Pirates Of The Caribbean: On Stranger Tides"
M = "The Chronicles Of Narnia"
M = "Prince Of Persia: The Sands Of Time"
...
```
Για παράδειγμα, το παραπάνω ερώτημα θα επιστρέφει αρκετά κοινές σε περιεχόμενο ταινίες με την ταινια "Pirates Of The Caribbean". Θα υπάρχουν και αντίστοιχα ερωτήματα που θα βρίσκουν λιγότερο όμοιες ταινίες. Ο δείκτης ομοιότητας των ταινιών είναι αυθαίρετος και μπορείτε να τον ορίσετε εσείς όπως θέλετε, αρκεί να υπάρχει κάποια λογική σύνδεση με τα δεδομένα που περιέχονται στο αρχείο movies_metadata.csv.

Συνεπώς η συνάρτηση που θα κάνει τις προτάσεις (recommendation) με είσοδο μια ταινία πρέπει να επιστρέφει-εκτυπώνει μια λίστα με τις προτεινόμενες ταινίες κατά φθίνουσα σειρά ομοιότητας.


###3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις - Αξιολογήσεις του χρήστη


Σε αυτό το σημείο θα εργαστείτε με τα αρχεία ratings τα όποια περιέχουν αξιολογήσεις (από 1 μέχρι 5) για τις παραπάνω ταινίες. Το προηγούμενο σύστημα συστάσεων προτείνει στον χρήστη ταινίες αποκλειστικά με βάση την ομοιότητά τους. Σε αυτό το σημείο θα γίνει μια αναβάθμιση του συστήματος έτσι ώστε να παράγονται καλύτερες  συστάσεις οι όποιες θα λαμβάνουν υπόψιν και τις προτιμήσεις του χρήστη, οι όποιες θα εξάγονται από τις αξιολογήσεις που έχει κάνει μέχρι στιγμής.

Η εκπαίδευση του recommender θα γίνεται ως εξής:

Για κάθε ταινία θα υπάρχει ένα score το όποιο αρχικά θε είναι ίσο με 0 και θα διαμορφώνεται από τις αξιολογήσεις κάθε user. Έτσι για έναν χρήστη με βάση τις αξιολογήσεις που υπάρχουν στο αρχείο train_ratings θα πρέπει:

1.	Για κάθε ταινία που έχει βαθμολογήσει να βρίσκονται οι κοινές ταινίες ανά κλίμακα και στο μέχρι τώρα σκορ κάθε παρόμοιας ταινίας θα προστίθεται ένα βάρος το όποιο θα μπορούσε να είναι το ποσοστό ομοιότητας της ταινίας (δηλαδή ένα βάρος για κάθε κλίμακα, αν δύο ταινίες μοιάζουν ενισχύουμε το βάρος που προσθέτουμε από το να μοιάζουν λιγότερο) επί τον βαθμό που έχει βάλει ο χρήστης για την αρχική ταινία (διαφορετικό είναι ο χρήστης να έχει βάλει 5/5 ή 1/5 σε μια ταινία από 3/5).

2. Στη συνέχεια, ανάλογα με το σκορ που έχει σχηματιστεί για κάθε ταινία, θα επιλέγεται αν αυτή θα μπορούσε να είναι προτεινόμενη για τον χρήστη ή όχι και θα μετράμε το πόσο καλά τα πήγε το σύστήμά μας με βάση ορισμένες μετρικές.

Η λογική πίσω από την παραπάνω διαδικασία είναι ότι παρόμοιες ταινίες θα έχουν ανάλογο βαθμό. Για παράδειγμα αν ένας χρήστης έχει αξιολογήσει αρκετές  ταινίες οι όποιες είναι sci-fiction με 5/5 τότε μια ταίνια sci-fiction την όποια δεν έχει δεί λογικά θα του αρέσει και θα έπρεπε να την προτείνουμε.

Μετά την εκπαίδευση του συστήματος σας καλείστε να δοκιμάσετε τον recommender που κατασκευάσατε στην πράξη. Για τον σκοπό αυτό θα φορτώσετε το αρχείο test_ratings.csv όπου περιέχονται οι αξιολογήσεις του ίδιου χρήστη για άλλες ταινίες. Το σύστημά σας πρέπει να προβλέπει αν μια ταινία θα πρέπει να προταθεί στον χρήστη. Μια ταίνια θεωρούμε οτι προτείνεται στον  χρήστη, αν έχει βαθμό μεγαλύτερο του 3 στο test_ratings. Συνεπώς για την επίβλεψη του συστήματός σας θα πρέπει για κάθε μια από τις ταινίες του αρχείου test_ratings.csv να προβλέψετε εάν θα αρέσει στο χρήστη ή όχι ώστε να την προτείνετε.

ΠΡΟΣΟΧΗ! Η "βαθμολογία" που θα υπολογίσετε για κάθε ταινία κατά το training δεν είναι απαραίτητα πρόβλεψη της βαθμολογίας που θα έβαζε ο χρήστης.

Στην συνέχεια, σε συνδιασμό με τις πραγματικές απαντήσεις του χρήστη θα αξιολογήσετε το σύστημά σας χρησιμοποιόντας τις μετρικές: precision, recall, f1 οι οποίες είναι οι πλέον γνωστές μετρικές και ευρέως χρησιμοποιούμενες τεχνικές για την επίβλεψη-μέτρηση απόδοσης ανάλογων συστημάτων.

1. Precision: Δείχνει πόσο ακριβές είναι το σύστημα. Υπολογίζει πόσα από τα στιγμιότυπα τα όποια προβλέψαμε ότι ανήκουν σε μια κλάση όντως ανήκουν σε αυτή. Η μετρική αυτή μας δίνει μια εικόνα σχετικά με τον αριθμό των ταινίων που προβλέψαμε ως προτεινόμενες ενώ δεν θα έπρεπε.

2. Recall: Υπολογίζει πόσα από τα στιγμιότυπα που ανήκουν σε μια κλάση (π.χ. προτεινόμενες ταινίες) προβλέφθηκαν σωστά.

3. F1: Είναι ένας μέσος μεταξύ των παραπάνω δυο μετρικών, έτσι ώστε να διατηρείται μια ισορροπία μεταξύ τους. Υπολογίζεται από την παρακάτω σχέση:

$$F_1=2\frac{Precision\times{Recall}}{Precision+Recall}$$

Οι παραπάνω συναρτήσεις παρέχονται από την βιβλιοθήκη scikit-learn.

Τέλος για την καλύτερη επίβλεψη του συστήματος σας μπορείτε να εκπαιδεύσετε  τον recommender σας με ένα υποσύνολο ταινίων από ελάχιστες, λίγες μέχρι και πολλές (π.χ. 3, 5, 10, 50, ...) για να μελετήσετε κατά πόσο σας βοηθούν οι επιπλέον αξιολογήσεις κάθε φορά (δηλαδή κατά πόσο βελτιώνονται οι παραπάνω μετρικές στο test set).

Έτσι π.χ. μπορείτε να εντοπίσετε περιπτώσεις όπως για παράδειγμα ότι με έναν recommender μπορεί να μην επιτυγχάνετε πολύ υψηλό σκορ όσο με άλλους, αλλά το βέλτιστο σκόρ σας επιτυγχάνεται πολύ γρηγόρα π.χ. με μόνο 10 ταινίες αντί 100. Έτσι για παράδειγμα αν ο αλγόριθμος τα πηγαίνει πολύ καλά για τρεις ταινίες και στη συνέχεια το σκορ βελτιώνεται ελάχιστα τότε αυτός ενδείκνυται για ένα σύστημα συστάσεων για νέους χρήστες όπου δεδομένου λίγων ταινιών ο αλγόριθμος είναι σε θέση να το να προτείνει καλές συστάσεις. Ενώ αν η καλύτερη απόδοση του αλγορίθμου σας είναι βέλτιστη με περισσότερες ταινίες π.χ. 50 τότε αυτός ο αλγόριθμος συστάσεων ενδείκνυται για παλιούς χρήστες με πολλές αξιολόγησεις.

---
## Κατασκευή Περιβάλλοντος Εργασίας
---
### Διάβασμα Αρχείων στο Colab (Μόνο για το Colab)
Αν η υλοποίηση γίνει στο google colab τότε μπορεί να χρησιμοποιηθεί το google drive ως file system. Για να γίνει Mount το google drive τρέχουμε τον παρακάτω κώδικα και κλικάρουμε στο link που θα μας εμφανιστεί.

```
from google.colab import drive
drive.mount('/content/drive')
import os
os.listdir('/gdrive/My Drive')
```
Έπειτα στην σελίδα που άνοιξε επιλέγουμε το mail μας και στο επόμενο παράθυρο που θα μας ανοίξει πατάμε Να επιτρέπεται. Στην συνέχεια αντιγράφουμε τον κωδικό που θα μας βγάλει και τον κάνουμε paste στο Input που έχει ανοίξει στο colab. Έτσι πλέον αν έχουμε ανεβάσει ένα αρχείο στο google drive μπορούμε να το βρούμε στην θέση:

```
movies_filename = '/gdrive/My Drive/' + movies_metadata.csv
```
Μπορούμε πλέον κανονικά να δουλέψουμε φτιάχνοντας φακέλους ή αρχεία και γενικότερα κάνοντας οτιδήποτε θα κάναμε αν ήμασταν τοπικά.

### Prolog μέσω Python

Το πακέτο που θα χρησιμοποιηθεί για την επικοινωνία Python και Prolog είναι το pyswip (https://pypi.org/project/pyswip/). Για να δουλέψει το Pyswip  χρειάζεται να υπάρχει το Swi-Prolog το όποιο αν δουλεύουμε τοπικά πρέπει να το εγκαταστήσουμε, ακολουθώντας αντίστοιχες οδηγίες στην σελίδα του εργαλείου. Για να γίνει του Swi-Prolog η εγκατάσταση στο Google Colab πρέπει να τρέξουμε τον παρακάτω κώδικα:

```
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog/stable
!sudo apt-get update
!sudo apt-get install swi-prolog
```
Σε κάποιο σημείο της εκτέλεσης εμφανίζεται ένα μήνυμα ότι πρέπει να πατήσουμε enter σε ένα input για να συνεχίσει η διαδικασία. Έπειτα από αυτό η εκτέλεση θα συνεχίσει χωρίς κάποιο πρόβλημα.

Τέλος πρέπει να εγκαταστήσουμε το pyswip (**όπου και να δουλεύουμε**) όπως παρακάτω:

```
!pip isntall pyswip
```


---
#Κώδικας για κατασκευή περιβάλλοντος εργασίας

##Μόνο για Google Colab

Κώδικας για να γίνει Mount to Google Drive

In [1]:
from google.colab import drive
drive.mount('/content/drive')

import os
path="drive/MyDrive/to_upload"

Mounted at /content/drive


In [2]:
%%capture
#install swi-prolog
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog
!sudo apt-get update
!sudo apt-get install swi-prolog
#install pyswip
!pip install pyswip


# **Μέρος 1: Μελέτη των Metadata, Δημιουργία κόσμου και των βασικών queries.**


## **1.1**

In [3]:
import pandas as pd
from pyswip import Prolog

In [4]:
# Η βιβλιοθήκη pandas είναι χρήσιμη για την εργασία με τέτοια δεδομένα
import pandas as pd
path = '/content/drive/MyDrive/to_upload/'
# Διάβασμα του αρχείου 'movie_metadata.csv'
data = pd.read_csv(path + "movies_metadata.csv")
#Στο csv υπαρχούν κελία με nan τιμές
#Στις θέσεις αυτές βάζουμε 'UNK' πράγμα που  κάνουμε με την παρακάτω συνάρτηση
data.fillna("UNK", inplace=True)
# Preview the first 5 lines of the loaded data
data.head()

<ipython-input-4-801f20c86fde>:8: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'UNK' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  data.fillna("UNK", inplace=True)


,Unnamed: 0,budget,genres,homepage,id,plot_keywords,language,original_title,overview,popularity,...,tagline,movie_title,vote_average,num_voted_users,title_year,country,director_name,actor_1_name,actor_2_name,actor_3_name
0,0,237000000,Action|Adventure|Fantasy|Science Fiction,http://www.avatarmovie.com/,19995,culture clash|future|space war|space colony|so...,English,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,...,Enter the World of Pandora.,Avatar,7.2,11800,2009.0,United States of America,James Cameron,Zoe Saldana,Sigourney Weaver,Stephen Lang
1,1,300000000,Adventure|Fantasy|Action,http://disney.go.com/disneypictures/pirates/,285,ocean|drug abuse|exotic island|east india trad...,English,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,...,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500,2007.0,United States of America,Gore Verbinski,Orlando Bloom,Keira Knightley,Stellan Skarsgård
2,2,245000000,Action|Adventure|Crime,http://www.sonypictures.com/movies/spectre/,206647,spy|based on novel|secret agent|sequel|mi6|bri...,Français,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,...,A Plan No One Escapes,Spectre,6.3,4466,2015.0,United Kingdom,Sam Mendes,Christoph Waltz,Léa Seydoux,Ralph Fiennes
3,3,250000000,Action|Crime|Drama|Thriller,http://www.thedarkknightrises.com/,49026,dc comics|crime fighter|terrorist|secret ident...,English,The Dark Knight Rises,Following the death of District Attorney Harve...,112.312950,...,The Legend Ends,The Dark Knight Rises,7.6,9106,2012.0,United States of America,Christopher Nolan,Michael Caine,Gary Oldman,Anne Hathaway
4,4,260000000,Action|Adventure|Science Fiction,http://movies.disney.com/john-carter,49529,based on novel|mars|medallion|space travel|pri...,English,John Carter,"John Carter is a war-weary, former military ca...",43.926995,...,"Lost in our world, found in another.",John Carter,6.1,2124,2012.0,United States of America,Andrew Stanton,Lynn Collins,Samantha Morton,Willem Dafoe


In [5]:
def clean_text(text):
  text = text.replace(u'\xa0', u'')
  text = text.replace(u"'", u'')
  text = text.replace("\\","")
  text = text.replace('"', '')
  text = text.replace('\n', ' ')
  text = text.strip()

  return text

Για να ξέρουμε τι πρέπει να βάλουμε στον κόσμο μας εκτελούμε το παρακάτω κελί και λαμβάνουμε τα δεδομένα που πρέπει να χρησιμοποιήσουμε(για την ακρίβεια την κατηγορία τους)

In [6]:
the_columns=data.columns
print(the_columns)

Index(['Unnamed: 0', 'budget', 'genres', 'homepage', 'id', 'plot_keywords',
       'language', 'original_title', 'overview', 'popularity',
       'production_companies', 'production_countries', 'release_date', 'gross',
       'duration', 'spoken_languages', 'status', 'tagline', 'movie_title',
       'vote_average', 'num_voted_users', 'title_year', 'country',
       'director_name', 'actor_1_name', 'actor_2_name', 'actor_3_name'],
      dtype='object')


In [7]:
import ast
# Create World
# Ορίζουμε τον κόσμο μας
prolog = Prolog()

# Για κάθε row του πίνακα φτιάχνουμε τα κατηγορήματα που θέλουμε να αποθηκέυσουμε
# αρχικά σε μια λίστα με το όνομα literals
literals = []
movie_score = {}

for row in data.itertuples(index=True, name='Pandas'):
  movie_title = clean_text(getattr(row, 'movie_title'))
  genre = clean_text(str(getattr(row, 'genres')))
  director_name = clean_text(getattr(row, 'director_name'))
  plot_keywords = clean_text(getattr(row, 'plot_keywords'))
  actor_1_name = clean_text(getattr(row, 'actor_1_name'))
  actor_2_name = clean_text(getattr(row, 'actor_2_name'))
  actor_3_name = clean_text(getattr(row, 'actor_3_name'))
  languages = clean_text(getattr(row, 'language'))
  production_companies = ast.literal_eval(getattr(row, 'production_companies'))   # String to list conversion
  production_countries = ast.literal_eval(getattr(row, 'production_countries'))
  title_year = clean_text(str(getattr(row, 'title_year')))

  # Some extra movie information
  rating = clean_text(str(getattr(row, 'vote_average')))
  duration = clean_text(str(getattr(row, 'duration')))
  gross = clean_text(str(getattr(row, 'gross')))
  popularity = clean_text(str(getattr(row, 'popularity')))

  if movie_title != 'UNK':
    # Gernes
    if genre != 'UNK':
      for g in genre.split("|"):
        literals.append("genre('"+ movie_title +"','"+ g +"')")
    # Director
    if director_name != 'UNK':
      literals.append("director('"+ movie_title +"','"+ director_name +"')")
    # Plot and Color
    bw = False
    if plot_keywords != 'UNK':
      for pk in plot_keywords.split("|"):
        if pk == "black and white":
          color = "blackwhite"
          literals.append("color('"+ movie_title +"','"+ color +"')")
          bw = True
        else:
          literals.append("keyword('"+ movie_title +"','"+ pk +"')")
      if not bw:
        color = "colored"
        literals.append("color('"+ movie_title +"','"+ color +"')")
    # Actors
    if actor_1_name != 'UNK':
      literals.append("actor('"+ movie_title +"','"+ actor_1_name +"')")
    if actor_2_name != 'UNK':
      literals.append("actor('"+ movie_title +"','"+ actor_2_name +"')")
    if actor_3_name != 'UNK':
      literals.append("actor('"+ movie_title +"','"+ actor_3_name +"')")
    # Language
    if languages != 'UNK':
      literals.append("language('"+ movie_title +"','"+ languages +"')")
    # Production Company
    if production_companies != "UNK":
      for company in production_companies:
        literals.append("company('"+ movie_title +"','"+ clean_text(company.get("name")) +"')")
    # Production Country
    if production_countries != "UNK":
      for country in production_countries:
        literals.append("country('"+ movie_title +"','"+ clean_text(country.get("name")) +"')")
    # Decade
    if title_year != 'UNK':
      year = float(title_year)
      a = int(year)%10
      decade = str(int(year - a + 10)) + 's'
      literals.append("decade('"+ movie_title +"','"+ decade +"')")
    # Stars
    if rating != 'UNK':
      if float(rating) >= 8.5:
        literals.append("stars('"+ movie_title +"','"+ "5" +"')")
      elif float(rating) < 8.5 and float(rating) >= 7.0:
        literals.append("stars('"+ movie_title +"','"+ "4" +"')")
      elif float(rating) < 7.0 and float(rating) >= 6.0:
        literals.append("stars('"+ movie_title +"','"+ "3" +"')")
      elif float(rating) < 6.0 and float(rating) >= 3.0:
        literals.append("stars('"+ movie_title +"','"+ "2" +"')")
      elif float(rating) < 3.0:
        literals.append("stars('"+ movie_title +"','"+ "1" +"')")
    # Duration
    if duration != 'UNK':
      if float(duration) >= 180.0:
        literals.append("duration('"+ movie_title +"','very long')")
      elif float(duration) < 180.0 and float(duration) >= 120.0:
        literals.append("duration('"+ movie_title +"','long')")
      elif float(duration) < 120.0 and float(duration) >= 60.0:
        literals.append("duration('"+ movie_title +"','normal')")
      elif float(duration) < 60.0:
        literals.append("duration('"+ movie_title +"','short')")
    # Gross
    if gross != 'UNK':
      if float(gross) >= 1000000000:
        literals.append("gross('"+ movie_title +"','"+ "billion" +"')")
      elif float(gross) < 1000000000 and float(gross) >= 750000000:
        literals.append("gross('"+ movie_title +"','"+ "expensive" +"')")
      elif float(gross) < 750000000 and float(gross) >= 350000000:
        literals.append("gross('"+ movie_title +"','"+ "normal" +"')")
      elif float(gross) < 350000000:
        literals.append("gross('"+ movie_title +"','"+ "cheap" +"')")
    # Popularity
    if popularity != 'UNK':
      if float(popularity) >= 150.000000:
        literals.append("popularity('"+ movie_title +"','"+ "blockbuster" +"')")
      if float(popularity) < 150.000000 and float(popularity) >= 90.000000:
        literals.append("popularity('"+ movie_title +"','"+ "popular" +"')")
      if float(popularity) < 90.000000 and float(popularity) >= 30.000000:
        literals.append("popularity('"+ movie_title +"','"+ "known" +"')")
      if float(popularity) < 30.000000:
        literals.append("popularity('"+ movie_title +"','"+ "unpopular" +"')")

# Η Prolog θέλει τα κατηγορήματά της με την σειρά
literals.sort()
for literal in literals:
  prolog.assertz(literal)

# Επίσης μπορούμε να κάνουμε consult ένα έτοιμο αρχείο στον κόσμο όπως παρακάτω
#prolog.consult(path + "db.pl")

In [8]:
 #prolog.assertz('(directed_by(X,Y) :- findall(M,director(M,X),Y))')

## **1.2**

In [9]:
#1 Five common genres
prolog.assertz('(common_genre_4(X, Y) :- genre(X, G1), genre(Y, G1), genre(X, G2), genre(Y, G2), genre(X, G3), genre(Y, G3), genre(X, G4), genre(Y, G4), G1\=G2, G1\=G3, G1\=G4, G2\=G3, G2\=G4, G3\=G4, X\=Y)')

In [10]:
#2 Three common genres
prolog.assertz('(common_genre_3(X, Y) :- genre(X, G1), genre(Y, G1), genre(X, G2), genre(Y, G2), genre(X, G3), genre(Y, G3), G1\=G2, G1\=G3, G2\=G3, X\=Y)')

In [11]:
#3 One same genre
prolog.assertz('(same_genre(X, Y) :- genre(X, G1), genre(Y, G1), X\=Y)')

In [12]:
#4 Same director
prolog.assertz('(common_director(X, Y) :- director(X, D), director(Y, D), X\=Y)')

In [13]:
#5 with 4 keywords matching
prolog.assertz('(common_plot_4(X, Y) :- keyword(X, K1), keyword(Y, K1), keyword(X, K2), keyword(Y, K2), keyword(X, K3), keyword(Y, K3), keyword(X, K4), keyword(Y, K4), K1\=K2, K1\=K3, K1\=K4, K2\=K3, K2\=K4, K3\=K4, X\=Y)')

In [14]:
#6 with 3 keywords matching
prolog.assertz('(common_plot_3(X, Y) :- keyword(X, K1), keyword(Y, K1), keyword(X, K2), keyword(Y, K2), keyword(X, K3), keyword(Y, K3), K1\=K2, K1\=K3, K2\=K3, X\=Y)')

In [15]:
#7 Three same main actors
prolog.assertz('(common_actors_3(X, Y) :- actor(X, A1), actor(Y, A1), actor(X, A2), actor(Y, A2), actor(X, A3), actor(Y, A3), A1\=A2, A1\=A3, A2\=A3, X\=Y)')

In [16]:
#8 Two same main actors
prolog.assertz('(common_actors_2(X, Y) :- actor(X, A1), actor(Y, A1), actor(X, A2), actor(Y, A2), A1\=A2, X\=Y)')

In [17]:
#9 One common actor
prolog.assertz('(common_actor(X, Y) :- actor(X, A1), actor(Y, A1), X\=Y)')

In [18]:
#10 Same language
prolog.assertz('same_language(E1, E2) :- language(E1, L), language(E2, L), E1 \= E2')

In [19]:
#11 Black or white
prolog.assertz('(common_colored(X, Y) :- color(X, C), color(Y, C), X\=Y)')

In [20]:
#12 common studio
prolog.assertz('common_studio(X, Y) :- company(X, Studio), company(Y, Studio), X \= Y')

In [21]:
#13 common production country
prolog.assertz('common_country(X, Y) :- country(X, Country), country(Y, Country), X \= Y')

In [22]:
#14 common decade
prolog.assertz('(common_decade(X, Y) :- decade(X, D), decade(Y, D), X\=Y)')

# **2ο Μέρος: Recommendation System με βάση μόνο τα χαρακτηριστικά των ταινιών.**

Στο σημείο αυτό με βάση τους κανόνες που κατασκευάστηκαν στο Μέρος 1 θα κατασκευαστούν κατηγορήματα για την ομοιότητα ταινιών. Παρακάτω δίνεται ένα μικρό παράδειγμα ενός κανόνα και πώς αυτός θα μπορούσε να γραφτεί μέσω του Pyswip. Επίσης όπως αναφέρεται και σε σχόλιο παραπάνω θα μπορεί να γραφτεί και μια βάση δεδομένων με τους κανόνες και να γίνει απευθείας consult.

Στο παρακάτω παράδειγμα το 5 και το 4 εκφράζουν την ομοιότητα των ταινιών π.χ. οι ταινίες που παράγονται μέσω του find_similar_5 είναι πιο όμοιες από αυτές που παράγονται μέσω του find_similar_4.


In [23]:
  #similarity 5
prolog.assertz('(find_similar_5(X, Y) :- common_genre_4(X, Y),common_plot_4(X, Y),common_actor_2(X, Y), same_language(X, Y); common_decade(X, Y); common_country(X, Y))')
  #similarity 4
prolog.assertz('(find_similar_4(X, Y) :- common_genre_3(X, Y), common_studio(X, Y), common_plot_3(X, Y))')
  #similarity 3
prolog.assertz('(find_similar_3(X, Y) :- common_genre_3(X, Y), common_country(X, Y); common_plot_3(X, Y))')
  #similarity 2
prolog.assertz('(find_similar_2(X, Y) :- same_genre(X, Y),common_studio(X,Y); common_director(X, Y))')
  #similarity 1
prolog.assertz('(find_similar_1(X, Y) :- same_genre(X, Y))')

Below are two code blocks for individual testing of each and every "find_similar_x" rule

In [24]:
def recommend_actors_similarity(movie):
    print(movie)
    s = set()
    for soln in prolog.query("find_similar_5('" + movie + "', Y)"):
      if soln['Y'] not in s:
        s.add(soln['Y'])
        print(soln['Y'])

In [25]:
recommend_actors_similarity('Avatar')

Avatar
(500) Days of Summer
102 Dalmatians
10th & Wolf
11:14
12 Rounds
13 Going on 30
1408
15 Minutes
16 Blocks
16 to Life
17 Again
2 Fast 2 Furious
2012
2046
21 Grams
21
25th Hour
27 Dresses
28 Days Later
28 Days
28 Weeks Later
2:13
3 Strikes
30 Days of Night
300
3000 Miles to Graceland
3:10 to Yuma
4 Months, 3 Weeks and 2 Days
40 Days and 40 Nights
50 First Dates
51 Birch Street
8 Mile
8 Women
88 Minutes
9
A Beautiful Mind
A Christmas Carol
A Cinderella Story
A Cock and Bull Story
A Dogs Breakfast
A Good Year
A Guy Thing
A History of Violence
A Home at the End of the World
A Knights Tale
A Lot Like Love
A Man Apart
A Mighty Heart
A Mighty Wind
A Perfect Getaway
A Prairie Home Companion
A Scanner Darkly
A Serious Man
A Shine of Rainbows
A Single Man
A Sound of Thunder
A Very Long Engagement
A Walk to Remember
A Woman, a Gun and a Noodle Shop
A.I. Artificial Intelligence
ATL
AVP: Alien vs. Predator
Abandon
Aberdeen
About Schmidt
About a Boy
Across the Universe
Adam Resurrected
Adam
Ada

In [26]:
def simple_recommender(movie):
    results = []
    for i in range(1, 6):
        s = set()
        try:
            q = prolog.query(f"find_similar_{i}('{movie}', M)")
            for soln in q:
                s.add(soln['M'])
            q.close()
        except:
            pass  # Ignore if rule is undefined or fails
        results.append(s)
    return results

In [27]:
list(simple_recommender('Avatar'))

[{'Tombstone',
  'The Great Escape',
  'Spice World',
  'The Dark Knight Rises',
  '1941',
  'Scooby-Doo',
  'Flight of the Intruder',
  'Spider-Man 2',
  'Special',
  'X-Men: Apocalypse',
  'Compadres',
  'Evil Dead II',
  'Time Changer',
  'Scott Pilgrim vs. the World',
  'Life of Pi',
  'Heroes of Dirt',
  'Starship Troopers',
  'Back to the Future',
  'Sky High',
  'The Darkest Hour',
  'Die Hard',
  'Ironclad',
  'The Crow',
  'Safe House',
  'The Hunger Games',
  'Bait',
  'The Illusionist',
  'Alvin and the Chipmunks',
  'The Twilight Saga: Eclipse',
  'Outbreak',
  'Miss Congeniality 2: Armed and Fabulous',
  'The Nutty Professor',
  'Furious 7',
  'Escape Plan',
  'Jack Ryan: Shadow Recruit',
  'The Adventures of Huck Finn',
  'The Four Feathers',
  'Decoys',
  'Braveheart',
  'The Siege',
  'Just Like Heaven',
  'Under the Skin',
  'Dances with Wolves',
  'Born Of War',
  'Sliding Doors',
  'Return to the Blue Lagoon',
  'Wasabi',
  'Last Holiday',
  'Godsend',
  'Black Novem

# **3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις-Αξιολογήσεις του χρήστη-Εκπαίδευση και Πρόβλεψη**


Αρχικά μελετάμε τις αξιολογήσεις κάθε χρήστη για να καταλάβουμε την δομή και τις πληροφορίες κάθε αρχείου.




In [28]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score
import numpy as np
import random

rating_weights = {0: -1, 1: -0.5, 2:0, 3:0, 4:0.5, 5:1}
score_weights = {i:i + 1 for i in range(5)} # ανάλογα με τα επίπεδα ομοιότητας που έχουν οριστεί στην simple_recommender

def train_recommender(ratings, rating_weights, score_weights, number_of_movies = 10):
    """
    Στην συνάρτηση αυτή μπορούμε να ορίζουμε ποιο υποσύνολο των αξιολογήσεων θα χρησιμοποιήσουμε για το train μαζί με τα βάρη ομοιότητας και σκορ
    Σε συνδυασμό με τον αριθμό των ταινιών που θέλουμε να χρησιμοποιήσουμε σαν σύνολο δεδομένων π.χ. 10 από τις 100 ή 3 από τις 100 κ.ο.κ
    Αν θέλουμε να χρησιμοποιήσουμε όλες τις ταινίες σαν training set τότε ορίζουμε το number_of_movies = - 1
    """

    if number_of_movies > len(ratings):
        number_of_movies = len(ratings)


    if number_of_movies != -1:
        indexes = random.sample(range(len(ratings)), number_of_movies)
        ratings = ratings.iloc[indexes]

    movie_score = {}
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        similar_movies = simple_recommender(movie)
        for i in range(5):
            for similar_movie in similar_movies[i]:
                if similar_movie not in movie_score:
                    movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[0]
                else:
                    movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[i]
    return movie_score

# Αυτό είναι ένα παράδειγμα για το πως θα μπορούσε να υλοποιήθει η predict.
# Έχουμε ορίσει ότι μια ταινία θα έπρεπε να είναι προτεινόμενη αν είχε σκορ > 0.
def predict_example(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score: #αν έχουμε σχηματίσει βαθμολογία για την ταινία αυτή
            pred.append(int(movie_score[movie] > 0)) #heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))# έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
            #η συνθήκη αυτή δεν μπορεί να αλλάξει
        else: #δεν μπορούμε να προτείνουμε κάτι για το όποιο δεν έχουμε σχηματίσει εικόνα
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred


def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics

Οι παραπάνω εκπαιδεύουν, τεστάρουν και μετρούν την απόδοσή του συστήματος στάσεων μας. Για την εκπαίδευση του συστήματος μπορούμε κάθε φορά να χρησιμοποιήσουμε ένα τυχαίο υποσύνολο του training set. Όμως είναι πιθανό το υποσύνολο των ταινιών αυτών να επηρεάζει τα αποτελέσματα στο training set. Για παράδειγμα από 3 ταινίες στις 10 μπορούν για ένα πείραμα τα αποτελέσματα μας αν είναι ίδια και αυτό να μην οφείλεται στο γεγονός ότι ο ταξινομητής μας τα πηγαίνει καλά στις 3 ταινίες αλλά στο γεγονός ότι οι υπόλοιπες 7 είναι τέτοιες ταινίες που δεν μας βοηθούν καθόλου και αν είχαμε επιλέξει διαφορετικές 3 ταινίες να τα πηγαίναμε χάλια. Οπότε προτείνουμε να τρέξετε κάθε φορά έναν αριθμό πειραμάτων για κάθε υποσύνολο ταινιών π.χ. 10 πειράματα με 3 ταινίες, 10 πειράματα για 20 ταινίες κ.ο.κ. και να κρατήσετε σαν τελικό σκορ το μέσο όρο όλων των πειραμάτων. Έτσι τα αποτελέσματα των αγγλικών σας θα είναι πιο αντικειμενικά.

In [29]:
train_ratings = pd.read_csv(path + "train_ratings.csv")
test_ratings = pd.read_csv(path + "test_ratings.csv")

In [30]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5083530752726622
recall: 0.9722222222222221
f1: 0.6669738329643412


In [31]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5095898875158076
recall: 0.9944444444444445
f1: 0.6738549106319782


In [32]:
# 50 movies
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 50)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

metrics_50_movies = {metric: 0 for metric in metrics[0].keys()}
for metric in metrics[0].keys():
    metrics_50_movies[metric] = np.mean([m[metric] for m in metrics])
    print (f"{metric}: {metrics_50_movies[metric]}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5106382978723404
recall: 1.0
f1: 0.676056338028169


In [33]:
# 100 movies
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 100)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

metrics_50_movies = {metric: 0 for metric in metrics[0].keys()}
for metric in metrics[0].keys():
    metrics_50_movies[metric] = np.mean([m[metric] for m in metrics])
    print (f"{metric}: {metrics_50_movies[metric]}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5106382978723404
recall: 1.0
f1: 0.676056338028169


# Σχόλια και Παρατηρήσεις


As shown by the  last code cells I can confidently say that the recommendation system that is created is pretty stable with a precision score of 50%. While it certainly is not the best the simple_recommender works with all the rules that were created, and were required for this exercise so i see no further optimisation beeing needed.The recall is steady on 1.0 validates the efficiency of the system that i build and thus it kind of "drives up" overall performance to around 67% as shown above,in almost every cell for the movies that we were supposed to test it for.
Overall the system is pretty solid and can achieve 1 in 2 movies to be considered interesting by the user, deeming it as having an acceptable performance.